app/code/Magento/Inventory/Model/SourceItem/Command/SourceItemsSave.php

```diff
@@ -25,27 +33,54 @@ class SourceItemsSave implements SourceItemsSaveInterface
      * @var LoggerInterface
      */
     private $logger;
+    /**
+     * @var ValidationResultFactory
+     */
+    private $validationResultFactory;
 
     /**
+     * @param ValidationResultFactory $validationResultFactory
+     * @param SourceItemValidatorInterface $sourceItemValidator
      * @param SaveMultiple $saveMultiple
      * @param LoggerInterface $logger
      */
     public function __construct(
+        ValidationResultFactory $validationResultFactory,
+        SourceItemValidatorInterface $sourceItemValidator,
         SaveMultiple $saveMultiple,
         LoggerInterface $logger
     ) {
+        $this->validationResultFactory = $validationResultFactory;
+        $this->sourceItemValidator = $sourceItemValidator;
         $this->saveMultiple = $saveMultiple;
         $this->logger = $logger;
     }
 
     /**
      * @inheritdoc
+     * @throws \Magento\Framework\Validation\ValidationException
      */
     public function execute(array $sourceItems)
     {
         if (empty($sourceItems)) {
             throw new InputException(__('Input data is empty'));
         }
+
+        // @todo should this be moved because of single responsibility
+        $errors = [];
+        foreach ($sourceItems as $sourceItem) {
+            $validationResult = $this->sourceItemValidator->validate($sourceItem);
+
+            if (!$validationResult->isValid()) {
+                $errors = array_merge($errors, $validationResult->getErrors());
```

Comment:
:exclamation:  Please Refactor **Performance Issue** 

 array_merge in foreach see :
https://github.com/kalessil/phpinspectionsea/blob/master/docs/performance.md#slow-array-function-used-in-loop